### PR TITLE
feat: read secrets from paths

### PR DESCRIPTION
**Pull request by nimish-ks**
## :mag: Overview
This PR implements a unified secrets management system for both frontend and backend, supporting Docker-style file-based secrets and environment variables. This enhancement streamlines the handling of sensitive configurations across Phase Console, particularly benefiting secrets and database credentials.

The implementation follows a consistent pattern across both stacks:
1. Check if `{key}_FILE` exists as an environment variable (e.g., `DATABASE_PASSWORD=/run/secrets/database_password`)
2. If present, read the secret from the specified file location
3. Otherwise, fall back to reading directly from environment variables

Docs PR: https://github.com/phasehq/docs/pull/125

### Backend Changes
- New `utils/secrets.py` module implementing the secrets provider pattern
- Support for Docker secret files and environment variables
- Debug logging for troubleshooting secret loading paths

### Frontend Changes
- New `utils/secretConfig.ts` module mirroring backend functionality
- Support for Docker secret files and environment variables
- Debug logging for troubleshooting secret loading paths

#### Backend (`utils/secrets.py`)
- Unified secrets management module
- Docker secrets support through file paths
- Debug logging when `DEBUG=True`

#### Frontend (`utils/secretConfig.ts`)
- TypeScript implementation of secrets management
- NextAuth.js integration for OAuth credentials
- Synchronous secret loading for critical paths

#### Backend Changes
```python
# Before
password = os.getenv('DATABASE_PASSWORD')

# After
from utils.secrets import get_secret
password = get_secret('DATABASE_PASSWORD')
```

#### Frontend Changes
```typescript
// Before
secret = process.env.NEXTAUTH_SECRET

// After
import { getSecret } from '@/utils/secretConfig'
secret = getSecret('NEXTAUTH_SECRET')
```

## :question: Open Questions
- NextAuth.js refused to start when the NEXTAUTH_SECRET is fetched via getSecret or even hard coded secret and not set via (process.env) in [frontend/pages/api/auth/[...nextauth].ts] @rohan-chaturvedi 

### Backend Testing
- Environment variable and file-based secret loading
- Priority handling between sources
- Test `PHASE_LICENSE = check_license(get_secret("PHASE_LICENSE_OFFLINE"))`
- Error cases and logging when DEBUG=True
- Edge cases (whitespace, empty files)

### Frontend Testing
- Secret initialization (SSO secrets)
- Edge cases (whitespace, empty files)
- Error cases and logging when DEBUG=True
- Jest test suite with fs mocking

## :sparkles: How to Test the Changes Locally

1. Create a development environment file:
```bash
cp .env.example .env.dev
```

If your .env.dev contains a secret that you are about to test the delivery via a path, please make sure to comment it out.
eg.

```fish
# GOOGLE_CLIENT_SECRET="your_google_secret"
☝🏻 
```

2. Set up required secrets in your local environment:
```bash
# Set OAuth secrets
export GITHUB_CLIENT_SECRET="your_github_secret"
export GOOGLE_CLIENT_SECRET="your_google_secret"
export GITLAB_CLIENT_SECRET="your_gitlab_secret"
export NEXTAUTH_SECRET="your_nextauth_secret"
export SECRET_KEY="your_django_secret"
```

3. Start the development environment:

Set the secrets you want to deliver via a path as secrets in your docker compose for example:

```yaml
services:
  nginx:
    container_name: phase-nginx-dev
    build:
      context: .
      dockerfile: ./nginx/Dockerfile
    restart: always
    ports:
      - 80:80
      - 443:443
    volumes:
      - ./nginx/default.conf:/etc/nginx/conf.d/default.conf:ro
    depends_on:
      - frontend
      - backend
    networks:
      - phase-net-dev

  frontend:
    container_name: phase-frontend-dev
    restart: unless-stopped
    depends_on:
      - backend
    build:
      context: ./frontend
      dockerfile: Dockerfile.dev
    volumes:
      - ./frontend/:/app/
      - /app/node_modules
    env_file: .env.dev
    environment:
      GITHUB_CLIENT_SECRET_FILE: /run/secrets/github_client_secret
      GOOGLE_CLIENT_SECRET_FILE: /run/secrets/google_client_secret
      GITLAB_CLIENT_SECRET_FILE: /run/secrets/gitlab_client_secret
    secrets:
      - nextauth_secret
      - github_client_secret
      - google_client_secret
      - gitlab_client_secret
    networks:
      - phase-net-dev

  backend:
    container_name: phase-backend-dev
    restart: unless-stopped
    depends_on:
      postgres:
        condition: service_healthy
      redis:
        condition: service_started
    build:
      context: ./backend
      dockerfile: Dockerfile.dev
    volumes:
      - ./backend/:/app/
    env_file: .env.dev
    environment:
      GITHUB_CLIENT_SECRET_FILE: /run/secrets/github_client_secret
      GOOGLE_CLIENT_SECRET_FILE: /run/secrets/google_client_secret
      GITLAB_CLIENT_SECRET_FILE: /run/secrets/gitlab_client_secret
     ☝🏻 
    secrets:
      - secret_key
      - github_client_secret
      - google_client_secret
      - gitlab_client_secret
     ☝🏻 
    networks:
      - phase-net-dev

  postgres:
    container_name: phase-postgres
    image: postgres:15.4-alpine3.17
    restart: always
    env_file:
      - .env.dev
    environment:
      POSTGRES_DB: ${DATABASE_NAME}
      POSTGRES_USER: ${DATABASE_USER}
      POSTGRES_PASSWORD: ${DATABASE_PASSWORD}
      POSTGRES_HOST_AUTH_METHOD: "trust"
    volumes:
      - phase-postgres-data-dev:/var/lib/postgresql/data
    networks:
      - phase-net-dev
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U ${DATABASE_USER} -d ${DATABASE_NAME}"]
      interval: 5s
      timeout: 5s
      retries: 5

  redis:
    container_name: phase-redis-dev
    image: redis:alpine3.19
    restart: always
    ports:
      - "6379:6379"
    networks:
      - phase-net-dev

  rqworker:
    container_name: phase-rqworker-dev
    restart: unless-stopped
    depends_on:
      - backend
      - redis
    build:
      context: ./backend
      dockerfile: Dockerfile.dev
    command: python manage.py rqworker default
    volumes:
      - ./backend/:/app/
    env_file: .env.dev
    secrets:
      - secret_key
    networks:
      - phase-net-dev

secrets:
  secret_key:
    environment: "SECRET_KEY"
  nextauth_secret:
    environment: "NEXTAUTH_SECRET"
  github_client_secret:
    environment: "GITHUB_CLIENT_SECRET"
  google_client_secret:
    environment: "GOOGLE_CLIENT_SECRET"
  gitlab_client_secret:
    environment: "GITLAB_CLIENT_SECRET"
☝🏻 

volumes:
  phase-postgres-data-dev:
    driver: local

networks:
  phase-net-dev:

```

```bash
docker compose -f dev-docker-compose.yml up -d
```

## :heavy_plus_sign: Additional Context
- Docker Secrets: https://docs.docker.com/compose/how-tos/use-secrets/
- NextAuth Configuration: https://next-auth.js.org/errors#no_secret

### :green_heart: Did You...
- [x] Ensure linting passes
- [x] Verify the app builds locally
- [x] Add comprehensive test coverage
- [x] Document the new functionality
- [x] Test NextAuth integration
- [x] Verify Docker secrets configuration
- [ ] Update dependencies (not required)
- [ ] Regenerate GraphQL schema (not required)

### CODE REVIEW

File: backend/backend/utils/secrets.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                if debug_mode:
+                    logger.debug(f"Loaded secret '{key}' from file: {file_path}")
+                return secret
+            except (PermissionError, OSError) as e:
+                if debug_mode:
+                    logger.debug(f"Failed to read secret file for '{key}': {e}")
+        elif debug_mode:
+            logger.debug(f"File path specified for '{key}' but file not found: {file_path}")
+            
+    secret = os.getenv(key, '')
```
Comment (rohan-chaturvedi) [inline backend/backend/utils/secrets.py]:
```suggestion
    secret = os.getenv(key, None)
```

File: frontend/utils/secretConfig.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                if (e instanceof Error) {
+                    console.debug(`[secrets] Failed to read secret file for '${key}': ${e.message}`)
+                }
+            }
+        }
+    } else if (debug && filePath) {
+        console.debug(`[secrets] File path specified for '${key}' but file not found: ${filePath}`)
+    }
+    
+    const secret = process.env[key] || ''
```
Comment (rohan-chaturvedi) [inline frontend/utils/secretConfig.ts]:
```suggestion
export function getSecret(key: string): string | undefined {
    const debug = process.env.DEBUG ? process.env.DEBUG === 'True' : false
    
    const fileEnvKey = `${key}_FILE`
    const filePath = process.env[fileEnvKey]
    
    if (filePath) {
        try {
            const secret = readFileSync(filePath, 'utf8').trim()
            if (debug) {
                console.debug(`[secrets] Loaded secret '${key}' from file: ${filePath}`)
            }
            return secret
        } catch (e) {
            if (debug) {
                if (e instanceof Error) {
                    console.debug(`[secrets] Failed to read secret file for '${key}': ${e.message}`)
                }
            }
        }
    } else if (debug && filePath) {
        console.debug(`[secrets] File path specified for '${key}' but file not found: ${filePath}`)
    }
    
    const secret = process.env[key] || undefined
```

**Review by rohan-chaturvedi — CHANGES_REQUESTED** [review]
Overall this looks great, just one change. Set the fall back value of secrets to `None` for python and `undefined` for ts. This should preserve application code behavior for env variable parsing when certain values are not provided.

**Review by rohan-chaturvedi — APPROVED** [review]
